### ISSUE TITLE
torch.compile(dynamic=True) specializes batch dimension for LayerNorm backward, causing recompiles

### ISSUE THREAD

**Issue by coolchameleon42** [pytorch/pytorch#172822]
### 🐛 Describe the bug

**Summary:**

When using `make_fx` with `tracing_mode='symbolic'` to trace a module containing `nn.LayerNorm` and `torch.autograd.grad(create_graph=True)`, and then compiling the resulting FX graph with `torch.compile(dynamic=True, fullgraph=True)`, the batch dimension (`dim 0`) gets specialized. This causes a new recompilation for every new batch size, triggered by a "guard size mismatch at index 0".

This behavior is specific to the presence of the `native_layer_norm_backward` operator in the graph. A forward-only `LayerNorm` graph compiles with a truly dynamic batch dimension as expected.

**Background:**
I am working on a machine learning force field project and aim to accelerate my model's training process using torch.compile. Briefly, the machine learning force field model reads distance vectors between atoms and outputs energy and forces, where the latter are derived by computing the gradient of energy with respect to the distance vectors. In other words, the forward() function of my model contains manually invoked autograd operations. Additionally, the loss function of my model consists of two components (energy and forces), which means that higher-order gradients need to be computed during the backward pass of training.

Meanwhile, the "batch_size" varies for each batch (since the number of atoms in each molecule is different), which requires that the code generated by compilation does not statically fix the dimensions of the input tensors.

To address this, I first traced my computation graph using make_fx(mode="symbolic"), then compiled the model with torch.compile(dynamic=True). However, I found that the compiled output still has statically fixed dimensions, leading to frequent recompilation when feeding inputs of different shapes. After analysis, I identified that the root cause of the recompilation is the backward operator of torch.nn.LayerNorm. A minimal reproducible example is provided below:

**Minimal Reproducer:**

```python
import os
import torch
import torch.nn as nn
from torch.fx import make_fx

# Set logs to observe guard failures and recompilations
os.environ["TORCH_LOGS"] = "dynamic,guards,recompiles"

class TinyDemo(nn.Module):
    """
    A simple model that applies LayerNorm and computes its gradient w.r.t. input.
    """
    def __init__(self, dim=64):
        super().__init__()
        self.dim = dim
        self.layer_norm = nn.LayerNorm(self.dim)

    def forward(self, x: torch.Tensor):
        x.requires_grad_(True)
        y = self.layer_norm(x)
        grad_outputs = [torch.ones_like(y)]
        # create_graph=True is essential to produce native_layer_norm_backward in the graph
        grad = torch.autograd.grad(
            [y],
            [x],
            grad_outputs=grad_outputs,
            create_graph=True,
            allow_unused=True
        )[0]
        return y, grad

# 1. Trace the model with symbolic tracing
model = TinyDemo(dim=64)
example_input = torch.randn(10, 64)
fx_model = make_fx(
    model,
    tracing_mode='symbolic',
    _allow_non_fake_inputs=True,
    _error_on_data_dependent_ops=True,
)(example_input)

print("--- FX Graph Code ---")
print(fx_model.code)

# 2. Compile the FX graph with dynamic=True
compiled_model = torch.compile(
    fx_model,
    dynamic=True,
    fullgraph=True
)

# 3. Run with varying batch sizes to trigger recompilations
print("\n--- Running with different batch sizes ---")
for n in [10, 20, 30, 40, 50, 60, 70, 80, 90]:
    print(f"Running with batch size N={n}")
    x = torch.randn(n, 64)
    try:
        y, g = compiled_model(x)
        print(f"  Success: y.shape={y.shape}, g.shape={g.shape}")
    except Exception as e:
        print(f"  Failed with error: {e}")

print("\n--- End of script ---")
```

[tl_out.zip](https://github.com/user-attachments/files/24736313/tl_out.zip)

### Error logs

**Expected Behavior:**

With `torch.compile(dynamic=True)`, the compiled model should handle a dynamic batch dimension (`N`) without recompiling. The first run (e.g., `N=10`) should produce a compiled artifact where the batch dimension is treated as a `SymInt`. Subsequent runs with different `N` values (20, 30, etc.) should reuse this artifact without triggering any guard failures related to the batch dimension.

**Actual Behavior:**

For each new batch size `N`, a recompilation is triggered. The logs show a series of guard failures indicating that the batch dimension was specialized to the previous value. For example, when running with `N=90` after running with `N=10` through `N=80`, the logs show:

> V0120... [__recompiles] Recompiling function forward...
> V0120... [__recompiles]     triggered by the following guard failure(s):
> V0120... [__recompiles]     - 0/7: tensor 'arg0_1' size mismatch at index 0. expected 80, actual 90
> V0120... [__recompiles]     - 0/6: tensor 'arg0_1' size mismatch at index 0. expected 70, actual 90
> ...
> V0120... [__recompiles]     - 0/0: tensor 'arg0_1' size mismatch at index 0. expected 10, actual 90

### Versions

PyTorch version: 2.9.1
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version: (Debian 12.2.0-14+deb12u1) 12.2.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.36

Python version: 3.12.4 (main, Dec 15 2025, 11:08:09) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.4.210.bsk.6-amd64-x86_64-with-glibc2.36
Is CUDA available: False
CUDA runtime version: 13.0.88

Nvidia driver version: Could not collect
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.16.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.16.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          128
On-line CPU(s) list:             0-127
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
Stepping:                        6
CPU(s) scaling MHz:              86%
CPU max MHz:                     3500.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4600.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       3 MiB (64 instances)
L1i cache:                       2 MiB (64 instances)
L2 cache:                        80 MiB (64 instances)
L3 cache:                        108 MiB (2 instances)
NUMA node(s):                    4
NUMA node0 CPU(s):               0-15,64-79
NUMA node1 CPU(s):               16-31,80-95
NUMA node2 CPU(s):               32-47,96-111
NUMA node3 CPU(s):               48-63,112-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] botorch==0.16.1
[pip3] gpytorch==1.15.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.5
[pip3] pytorch-lightning==2.6.0
[pip3] torch==2.9.1
[pip3] torch-runstats==0.2.0
[pip3] torchaudio==2.9.1+cu130
[pip3] torchmetrics==1.8.2
[pip3] torchtnt==0.2.4
[pip3] torchvision==0.24.1+cu130
[pip3] triton==3.5.1
[conda] Could not collect

cc @chauhang @penguinwu @ezyang @bobrenjc93

**Comment by laithsakka** [comment]
> > FunctionsManual.cpp
> 
> how did you identify "const int64_t M = ", it seems not directly shown in the log even with TORCH_LOGS=+dynamic TORCHDYNAMO_EXTENDED_DEBUG_CPP=1 TORCHDYNAMO_EXTENDED_DEBUG_GUARD_ADDED="Eq(s26, 50)"

the logs point to the function,  so then I can guess, if you build with debug symbols you will also get more detailed info
